### PR TITLE
ADD Dockerfile to enable dev with the latest Ubuntu and Docker Containers

### PR DESCRIPTION
**Pull request by jmcanterafonseca**
This is a work in progress. Please do not review yet. Early feedback welcome.

### CODE REVIEW

File: docker/dev/ubuntu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Develop Orion Context Broker Ubuntu-Latest
+FROM ubuntu:latest
```
Comment (LeandroGuillen) [inline docker/dev/ubuntu/Dockerfile]: Why Ubuntu and not CentOS?

File: docker/dev/ubuntu/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    wget https://github.com/mongodb/mongo-cxx-driver/archive/legacy-1.0.2.tar.gz && \
+    tar xfvz legacy-1.0.2.tar.gz && \
+    cd mongo-cxx-driver-legacy-1.0.2 && \
+    scons && \
+    scons install --prefix=/usr/local && \
+    cd .. && \
+    rm -R mongo-cxx-driver-legacy-1.0.2 && \
+    echo cd /root/src >> /root/.bashrc
+
+EXPOSE 1026
```
Comment (LeandroGuillen) [inline docker/dev/ubuntu/Dockerfile]: I understand this docker is for developing Orion. You will need to create a data volume in this dockerfile pointing to the root of the project tree.

File: docker/dev/ubuntu/scripts/examples/mongo.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+docker --tlsverify=false run -d --name mongoOrion mongo:2.6.10
```
Comment (LeandroGuillen) [inline docker/dev/ubuntu/scripts/examples/mongo.sh]: I woud suggest a docker compose file instead of this one. Or at least in addition to this one.

File: docker/dev/ubuntu/Dockerfile (hunk 1 above)
Comment (jmcanterafonseca) [inline docker/dev/ubuntu/Dockerfile]: yes, we are going to need the CentOS one as well. Maybe you can contribute it, taking as input the work done at https://registry.hub.docker.com/u/bitergia/fiware-orion/dockerfile/

File: docker/dev/ubuntu/Dockerfile (hunk 2 above)
Comment (jmcanterafonseca) [inline docker/dev/ubuntu/Dockerfile]: yes, that's the idea to keep the codebase outside the container in order to be able to edit source code locally

File: docker/dev/ubuntu/scripts/examples/mongo.sh (hunk 3 above)
Comment (jmcanterafonseca) [inline docker/dev/ubuntu/scripts/examples/mongo.sh]: oh yes, we need the composer as well

thanks for the feedback!

File: docker/dev/ubuntu/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# Develop Orion Context Broker Ubuntu-Latest
+FROM ubuntu:latest
+
+MAINTAINER FIWARE Orion Context Broker Team. Telefónica I+D
+
+RUN \
+    apt-get update && \
```
Comment (fgalan) [inline docker/dev/ubuntu/Dockerfile]: This script looks a bit obscure for people not used to the Orion build process. Maybe introducing comments in the proper places could help, e.g:

```
# Installing Orion build dependencies
...
# Building mongo driver
...
```

File: test/functionalTest/harnessFunctions.sh [hunk 5]
```diff
@@ -94,32 +94,45 @@ function dMsg()
 #
 function dbInit()
 {
-  role=$1  
-  db=$1
+  role=$1
+  
+  host="${CB_DATABASE_HOST}"
```
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: Why did you need to change the way in which some harness functions are implemented? Or, in other words, which aspect in the current way of working make them incompatible with the docker stuff.

File: docker/dev/ubuntu/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# Develop Orion Context Broker Ubuntu-Latest
```
Comment (fgalan) [inline docker/dev/ubuntu/Dockerfile]: As general comment, I think that some piece of documentation (i.e. an .md file) explaining how to use the dokcer stuff would be needed, in a similar way a .md file explaining the valgrand stuff is already in the repository  (see https://github.com/telefonicaid/fiware-orion/blob/develop/doc/manuals/vagrant.md)

File: docker/dev/ubuntu/Dockerfile (hunk 1 above)
Comment (fgalan) [inline docker/dev/ubuntu/Dockerfile]: I agree in that we should provide also CentOS 6.x (it would be a bit weird to include Ubuntu, which is not official, before including the official one). How complex is to create a docker file for CentOS 6.x?

File: test/functionalTest/harnessFunctions.sh (hunk 5 above)
Comment (LeandroGuillen) [inline test/functionalTest/harnessFunctions.sh]: It seems to me that this is so that the MongoDB instance can be on a host different than localhost for the harness tests. Which is the case when running docker.

File: docker/dev/ubuntu/Dockerfile (hunk 1 above)
Comment (dmoranj) [inline docker/dev/ubuntu/Dockerfile]: Change `FROM ubuntu:latest` to `FROM centos:6` and it's done. I don't think we should provide different versions of Context Broker for different SOs, as any Docker image should work on any Host SO without changes so, what do we have to gain having multiple images? Docker images should be black boxes of the applications, so there's no point, IMHO, in creating multiple flavours.

Comment (fgalan) [inline docker/dev/ubuntu/Dockerfile]: Not sure if it so simple... note that for instance the `apt-get update` in `RUN`will not work in CentOS.

Comment (LeandroGuillen) [inline docker/dev/ubuntu/Dockerfile]: I agree with Dani. But this image I understand is for development of Orion, not to deploy an Orion to use. I guess this is for folks who want to use other OS than CentOS for development. That's why I think the CentOS one should be preferred.

Then, for deploying Orion or just trying it this image wouldn't make much sense.

Comment (LeandroGuillen) [inline docker/dev/ubuntu/Dockerfile]: Orion docker based on CentOS 6 in https://github.com/telefonicaid/fiware-orion/pull/1059

File: docker/dev/ubuntu/Dockerfile (hunk 6 above)
Comment (fgalan) [inline docker/dev/ubuntu/Dockerfile]: PR https://github.com/telefonicaid/fiware-orion/pull/1059 contains a very good .md document that could be used as template for the Ubuntu case.

File: docker/dev/ubuntu/Dockerfile (hunk 1 above)
Comment (fgalan) [inline docker/dev/ubuntu/Dockerfile]: NTC from my side.

File: docker/dev/ubuntu/scripts/examples/orion-dev.sh [hunk 7]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+source $HOME/.bashrc
+docker --tlsverify=false run -p 1026:1026 --name orionDev --link mongoOrion:mongoOrion -v /Users/jmcf/work/develop/fiware-orion:/root/src -t -i orion-dev:latest /bin/bash
```
Comment (fgalan) [inline docker/dev/ubuntu/scripts/examples/orion-dev.sh]: `/Users/jmcf/work/develop/fiware-orion` is user-specific. Should be replaced in the documention, e.g. using some generic like  `/path/to/fiware-orion` or something like that.

**Comment by fgalan** [comment]
Taking into account how the docker requirements have evolved (and the merging of PR https://github.com/telefonicaid/fiware-orion/pull/1069 implementing such requirements), maybe this PR should be closed?

**Comment by jmcanterafonseca** [comment]
closing this. Non-official docker corresponding to this PR can be found at 

https://registry.hub.docker.com/u/jmcanterafonseca/fiware-orion-dev/
